plugins/rule_engines/src/metadata_guard.cpp

```diff
@@ -0,0 +1,368 @@
+#include "irods/irods_at_scope_exit.hpp"
+#include "irods/irods_get_full_path_for_config_file.hpp"
+#include "irods/irods_logger.hpp"
+#include "irods/irods_plugin_context.hpp"
+#include "irods/irods_re_plugin.hpp"
+#include "irods/irods_re_ruleexistshelper.hpp"
+#include "irods/irods_re_serialization.hpp"
+#include "irods/irods_rs_comm_query.hpp"
+#include "irods/irods_state_table.h"
+#include "irods/rodsError.h"
+#include "irods/rodsErrorTable.h"
+#include "irods/scoped_privileged_client.hpp"
+
+#define IRODS_QUERY_ENABLE_SERVER_SIDE_API
+#include "irods/irods_query.hpp"
+
+#define IRODS_USER_ADMINISTRATION_ENABLE_SERVER_SIDE_API
+#include "irods/user_administration.hpp"
+
+#include <fmt/format.h>
+#include <nlohmann/json.hpp>
+
+#include <boost/algorithm/string/predicate.hpp>
+#include <boost/any.hpp>
+
+#include <algorithm>
+#include <array>
+#include <optional>
+#include <string>
+#include <unordered_map>
+
+namespace
+{
+    // clang-format off
+    using log_re = irods::experimental::log::rule_engine;
+    using json   = nlohmann::json;
+    // clang-format on
+
+    auto get_rei(irods::callback& _effect_handler) -> ruleExecInfo_t&
+    {
+        ruleExecInfo_t* rei{};
+        irods::error result{_effect_handler("unsafe_ms_ctx", &rei)};
+
+        if (!result.ok()) {
+            THROW(result.code(), "Failed to get rule execution info");
+        }
+
+        return *rei;
+    }
+
+    auto load_plugin_config(ruleExecInfo_t& _rei) -> std::optional<json>
+    {
+        // Must elevate privileges so that the configuration can be retrieved.
+        // Users who aren't administrators cannot retrieve metadata they don't own.
+        irods::experimental::scoped_privileged_client spc{*_rei.rsComm};
+
+        const auto gql = fmt::format("select META_COLL_ATTR_VALUE "
+                                     "where META_COLL_ATTR_NAME = 'irods::metadata_guard' and COLL_NAME = '/{}'",
+                                     _rei.rsComm->myEnv.rodsZone);
+
+        if (irods::query q{_rei.rsComm, gql}; q.size() > 0) {
+            try {
+                return json::parse(q.front()[0]);
+            }
+            catch (const json::exception&) {
+                const char* msg = "Cannot parse Rule Engine Plugin configuration";
```

Comment:
for your consideration...

```suggestion
                const char* msg = "Cannot parse Metadata Guard configuration";
```